### PR TITLE
Sealed block importing and propagation optimization

### LINKED ISSUE #773: Mined block should be short-cutted into the chain
Currently just gets pushed into BlockQueue like an untrusted block, but we know it's good so the JournalDB should be committed directly and the info pushed into the blockchain DB.

```
// TODO: commit DB from `sealed.drain` and make a VerifiedBlock to skip running the transactions twice.
```

### LINKED ISSUE #1468: New block mined log output
Currently parity prints the following message when a new block is mined locally:

2016-06-28 06:39:13 New block mined, hash: xxx....xxx

Could you please include the tentative block number or the full hash of the mined block in the output so that it easier to find the block later on via a blockchain explorer?

Thanks!
All the best, Peter

### LINKED ISSUE #1473: Send block to all peers if we mined it

### PR DESCRIPTION
**Pull request by arkpar**
Fixes #773 
Fixes #1473
Fixes #1468

### CODE REVIEW

File: ethcore/src/client/client.rs [hunk 1]
```diff
@@ -259,12 +259,12 @@ impl Client {
...
-			for hash in route.enacted {
-				map.insert(hash, true);
+		let map = import_results.iter().fold(HashMap::new(), |mut map, route| {
+			for hash in &route.enacted {
+				map.insert(hash.clone(), true);
 			}
-			for hash in route.retracted {
-				map.insert(hash, false);
+			for hash in &route.retracted {
+				map.insert(hash.clone(), false);
```
Comment (rphmeier) [inline ethcore/src/client/client.rs]: could drastically simplify this function as

```rust
let (mut enacted, mut retracted) = (Vec::new(), Vec::new());
for route in import_results {
    enacted.extend(route.enacted.iter().cloned());
    retracted.extend(route.retracted.iter().cloned());
}

(enacted, retracted)
```

simpler, fewer allocations, shorter

Comment (tomusdrw) [inline ethcore/src/client/client.rs]: It's not the same - you can can have he same blocks in both `enacted` and `retracted` and we need to figure out what its the last state.

Comment (rphmeier) [inline ethcore/src/client/client.rs]: ah, yeah. i should have read the comment...

File: ethcore/src/client/client.rs [hunk 2]
```diff
@@ -351,7 +325,7 @@ impl Client {
 
 		{
 			if !imported_blocks.is_empty() && self.block_queue.queue_info().is_empty() {
-				let (enacted, retracted) = self.calculate_enacted_retracted(import_results);
+				let (enacted, retracted) = self.calculate_enacted_retracted(&import_results);
```
Comment (tomusdrw) [inline ethcore/src/client/client.rs]: Why `import_results` is not moved anymore?

Comment (tomusdrw) [inline ethcore/src/client/client.rs]: Ah, ok I see it now.

File: ethcore/src/service.rs [hunk 3]
```diff
@@ -36,6 +36,8 @@ pub enum SyncMessage {
 		retracted: Vec<H256>,
 		/// Hashes of blocks that are now included in cannonical chain
 		enacted: Vec<H256>,
+		/// Hashes of blocks that are sealed by this node
+		sealed: Vec<H256>,
```
Comment (tomusdrw) [inline ethcore/src/service.rs]: `Option<H256>` should be enough here, no? Could simplify couple of other places. But I'm good with `Vec` for constistency

**Pull request merged**